Allow deleting local replay file from disk.

fultonm opened the pull request:
Addresses #1227 

There are some new strings to translate:

- delete = Delete
- replay.deleteTooltip = Delete local replay file
- replay.delete = Delete Replay
- replay.deleteLocal = Delete local replay
- replay.deleteWarn = This action will permanently delete the replay from your hard drive.

Let me know what you think!

fultonm committed:
Allow deleting local replay file from disk.

Sheikah45 commented on src/main/java/com/faforever/client/main/event/DeleteLocalReplayEvent.java:
```diff
@@ -0,0 +1,12 @@
+import javafx.event.Event;
+import javafx.event.EventType;
+
+public class DeleteLocalReplayEvent extends Event {
+  public static final EventType<DeleteLocalReplayEvent> DELETE_LOCAL_REPLAY_EVENT_TYPE = new EventType<>(ANY);
+
+  public DeleteLocalReplayEvent() {
+    super(DELETE_LOCAL_REPLAY_EVENT_TYPE);
+  }
+}
```
For events we actually use the Google EventBus if you search for it in the source you should see a few examples. This allows arbitrary objects to be events

fultonm replied:
It seems I need to change a several classes constructors to have access to that EventBus. Does that sound correct or is there a better way to inject this dependency into LocalReplayVaultController? Any hint appreciated :)

fultonm replied:
@Sheikah45 when you get a chance, not sure if you saw my reply buried in here

Sheikah45 replied:
Yes you should add the eventbus to the constructor

Sheikah45 commented on src/main/java/com/faforever/client/replay/LocalReplayVaultController.java:
```diff
@@ -50,10 +51,15 @@ public void initialize() {
     replayDetailController.getRoot().requestFocus();
   }
 
+  private void onLocalReplayDeleted() {
+    onPageChange(searchController.getLastSearchConfig(), false);
+  }
```
Will want to convert this to use the google eventBus and so will need to annotate this method with @ Subscribe

Sheikah45 commented on src/main/java/com/faforever/client/replay/LocalReplayVaultController.java:
```diff
@@ -50,10 +51,15 @@ public void initialize() {
   protected void onDisplayDetails(ReplayBean replay) {
     JavaFxUtil.assertApplicationThread();
     replayDetailController.setReplay(replay);
+    replayDetailController.getRoot().addEventHandler(DeleteLocalReplayEvent.DELETE_LOCAL_REPLAY_EVENT_TYPE, event -> onLocalReplayDeleted());
```
Use google eventbus, might need to be added as a dependency through spring injection and also registration added in the initialize function

Sheikah45 commented on src/main/java/com/faforever/client/replay/ReplayDetailController.java:
```diff
@@ -425,6 +434,20 @@ public void onReport() {
+
+  private void deleteReplay() {
+    replayService.deleteReplayFile(replay.getReplayFile());
+    getRoot().fireEvent(new DeleteLocalReplayEvent());
```
use `eventbus.post`

Sheikah45 commented on src/main/java/com/faforever/client/replay/ReplayService.java:
```diff
@@ -266,6 +267,14 @@ private void moveCorruptedReplayFile(Path replayFile) {
+    try {
+      delete(replayFile);
+    } catch (IOException e) {
+      log.error("Failed to delete local replay file {}", replayFile, e);
```
probably want to add a user notification about the failure here

Sheikah45 commented on src/main/resources/i18n/messages.properties:
```diff
@@ -766,12 +767,16 @@ replay.replayDurationTooltip = Replay length
 replay.ratingTooltip = Average rating
 replay.qualityTooltip = Quality
+replay.deleteTooltip = Delete local replay file
+replay.delete = Delete Replay
+replay.deleteLocal = Delete local replay
```
why delete and deleteLocal I would think we only need one

fultonm replied:
`replay.delete` is used as a label for the Delete Replay button, while `replay.deleteLocal` is used as part of the confirmation notification title. I updated the translation entry keys to be more descriptive of their purpose. To me, the difference in capitalization and verbosity make sense in each of their context. Let me know if we need to minimize the number of translations.

Sheikah45 reviewed (changes requested):
Looking good. One note that you will see throughout is that we actuall use the google event bus for publsihing and subscribing to events. So if you convert to using that we should be in a good place.

Also as a thought I wonder if a delete button would fit on the local replay card as well.

fultonm committed:
Use colorful buttons on replay cards. Use Google event bus for delete replay message passing.

fultonm committed:
Use colorful buttons on replay cards. Use Google event bus for delete replay message passing.

Sheikah45 commented on src/main/java/com/faforever/client/main/event/DeleteLocalReplayEvent.java:
```diff
@@ -0,0 +1,12 @@
+@RequiredArgsConstructor
+public class DeleteLocalReplayEvent {
+  @Getter
+  private final Path replayFile;
```
```suggestion
@Value
public class DeleteLocalReplayEvent {
  Path replayFile;
```

Use @ Value here as it does everything you need with one annotation

Sheikah45 commented on src/main/java/com/faforever/client/replay/ReplayCardController.java:
```diff
@@ -75,7 +86,10 @@ public void setReplay(ReplayBean replay) {
-
+    if (replay.getReplayFile() != null) {
+      deleteButton.setManaged(true);
+      deleteButton.setVisible(true);
```
Instead of setting managed and visible true create an initialize function that uses the JavafxUtil.bindManagedToVisible() method so that we only need to set the visible flag as we do in other places

Sheikah45 commented on src/main/java/com/faforever/client/replay/ReplayDetailController.java:
```diff
@@ -277,6 +286,8 @@ public void setReplay(ReplayBean replay) {
       optionsTable.setItems(replay.getGameOptions());
       moreInformationPane.setVisible(true);
+      deleteButton.setVisible(true);
+      deleteButton.setManaged(true);
```
Same bindManagedToVisible comment as above. Here you can can just add it to the already existing call

fultonm commented:
At the top of ReplayController.java, here is a todo: `// TODO change to spring event bus`

Sheikah45 reviewed (changes requested):
Looking good just a few changes

fultonm committed:
Update button visible syntax and annotations

fultonm commented:
> At the top of ReplayController.java, here is a todo: `// TODO change to spring event bus`

@Sheikah45 should this be resolved or is it OK?

Sheikah45 commented:
Yes that can be removed everything was already converted to the event bus

fultonm committed:
Remove event bus todo

Sheikah45 commented on src/main/java/com/faforever/client/replay/ReplayCardController.java:
```diff
@@ -75,7 +86,10 @@ public void setReplay(ReplayBean replay) {
       onMapLabel.setText(i18n.get("game.onUnknownMap"));
     }
-
+    JavaFxUtil.bindManagedToVisible(deleteButton);
```
This should be done in an initialize function as it only needs to be done once

Sheikah45 commented on src/main/java/com/faforever/client/replay/ReplayCardController.java:
```diff
@@ -75,7 +86,10 @@ public void setReplay(ReplayBean replay) {
+    JavaFxUtil.bindManagedToVisible(deleteButton);
+    if (replay.getReplayFile() != null) {
+      deleteButton.setVisible(true);
+    }
```
```suggestion
  deleteButton.setVisible(replay.getReplayFile() != null);
```

Sheikah45 commented on src/main/resources/theme/vault/replay/replay_card.fxml:
```diff
@@ -84,11 +84,24 @@
+                visible="false" managed="false"
+                onAction="#onDeleteButtonClicked"
+                styleClass="delete-button">
+            <graphic>
+                <Region styleClass="icon,delete-icon"/>
+            </graphic>
```
This should be the same delete icon as the ones used for the maps and mods delete buttons

fultonm replied:
Updated this one to `uninstall-icon` like the maps and mods icons.

Sheikah45 commented on src/main/resources/theme/style.css:
```diff
@@ -19,6 +19,45 @@
+    -fx-background-color: derive(-fx-delete-color, -60%);
+}
+
+/**.main-navigation-button:highlighted {
+  -fx-background-color: -fx-accent;
+}
+
+.main-navigation-button:hover {
+  -fx-border-color: -fx-selection-bar-non-focused;
+  -fx-background: #FFFFFF12;
+}
+
+.main-navigation-button:armed, .main-navigation-button:selected, .main-navigation-button:focused:selected {
+  -fx-background-color: #FFFFFF33;
+  -fx-border-color: -fx-accent;
+}*//
```
Remove the commented out code

fultonm replied:
Oops, I was using this to template the delete button icon styling.

fultonm committed:
Standardize replay, map, and mod buttons. Initialize field properties in ReplayCardController.

fultonm commented:
I think the buttons on the replay, maps, and mods section should match in style as much as possible, so I added color to those as well.

- The colored buttons will highlight the shortcut to perform these actions directly from the card without the need to navigate into the detail view.
- Blue and Red make it clear what the buttons do without needing to recognize and process the meaning of the icon (less brain cells.)
- I think it will encourage players to actually use these buttons more.

Sheikah45 commented on src/main/java/com/faforever/client/replay/ReplayCardController.java:
```diff
@@ -160,4 +176,19 @@ public void onShowReplayDetail() {
+
+  public void onDeleteButtonClicked() {
+    notificationService.addNotification(new ImmediateNotification(
+      String.format("%s %s",
+          i18n.get("replay.deleteNotification.heading"),
+          replay.getTitle()),
```
Rather than use string.format you should add the argument to the string in the messages.properties file with the placeholder {0} then pass the replay title as an argument to the i18n.get call. This avoids the inherent locale dependency of String.format
```suggestion
          i18n.get("replay.deleteNotification.heading", replay.getTitle())
```

fultonm replied:
Nice, didn't realize this was a thing

Sheikah45 commented on src/main/java/com/faforever/client/replay/ReplayCardController.java:
```diff
@@ -160,4 +176,19 @@ public void onShowReplayDetail() {
+      String.format("%s %s",
+          i18n.get("replay.deleteNotification.heading"),
+          replay.getTitle()),
+      i18n.get("replay.deleteNotification.warning"),
+      Severity.INFO, Arrays.asList(
```
If the severity is just info then the message shouldn't be call warning

Sheikah45 commented on src/main/java/com/faforever/client/replay/ReplayDetailController.java:
```diff
@@ -425,6 +435,23 @@ public void onReport() {
 
+  public void onDeleteButtonClicked() {
+    notificationService.addNotification(new ImmediateNotification(
+      String.format("%s %s",
+        i18n.get("replay.deleteNotification.heading"),
+        replay.getTitle()),
```
Same i18n comment as above

Sheikah45 commented on src/main/java/com/faforever/client/replay/ReplayDetailController.java:
```diff
@@ -425,6 +435,23 @@ public void onReport() {
+      String.format("%s %s",
+        i18n.get("replay.deleteNotification.heading"),
+        replay.getTitle()),
+        i18n.get("replay.deleteNotification.warning"),
+        Severity.INFO, Arrays.asList(
```
same naming comment

Sheikah45 reviewed (changes requested):
Looking good just a few small changes left

fultonm committed:
Use i18n formatting instead of String.format.
Add event bus import to MapVaultControllerTest for test framework injection.
Merge recent develop branch changes to this branch.

Sheikah45 commented:
Hmm it looks like some of the rebasing onto develop didn't happen properly as some commits are duplicated here

fultonm commented:
> Hmm it looks like some of the rebasing onto develop didn't happen properly as some commits are duplicated here

I used merge instead of rebase, so I had to back up it up and reorder some commits and rebase instead. I force pushed to fix the history

Sheikah45 commented:
Although it looks like some tests are failing

Sheikah45 committed:
Fix test

Sheikah45 committed:
Fix test

Sheikah45 committed:
optimize imports

Sheikah45 committed:
explicit Files calls
